sway-core/src/type_system/info.rs

```diff
@@ -1156,6 +1156,43 @@ impl TypeInfo {
         matches!(self, TypeInfo::Tuple(_))
     }
 
+    pub fn contains_never(&self, engines: &Engines) -> bool {
+        match self {
+            TypeInfo::Never => true,
+            TypeInfo::Tuple(type_arguments) => type_arguments
+                .iter()
+                .any(|t| engines.te().get(t.type_id).contains_never(engines)),
+            TypeInfo::Array(type_argument, _) => engines
+                .te()
+                .get(type_argument.type_id)
+                .contains_never(engines),
+            TypeInfo::Unknown
+            | TypeInfo::UnknownGeneric { .. }
+            | TypeInfo::Placeholder(_)
+            | TypeInfo::TypeParam(_)
+            | TypeInfo::StringSlice
+            | TypeInfo::StringArray(_)
+            | TypeInfo::UnsignedInteger(_)
+            | TypeInfo::Enum(_)
+            | TypeInfo::Struct(_)
```

Comment:
Is this completely correct? We can have uninhabited structs and enums. E.g., this code compiles without errors, as it should:
```Sway
script;

struct S {
    x: !,
}

fn main() -> u64 {
    let s = S { x: { return 0u64; } };    
    0
}
```